[lldb] Fix evaluating expressions without JIT in an object context

igorkudrin opened the pull request:
If a server does not support allocating memory in an inferior process or when debugging a core file, evaluating an expression in the context of a value object results in an error:

```
error: <lldb wrapper prefix>:43:1: use of undeclared identifier '$__lldb_class'
   43 | $__lldb_class::$__lldb_expr(void *$__lldb_arg)
      | ^
```

Such expressions require a live address to be stored in the value object. However, `EntityResultVariable::Dematerialize()` only sets `ret->m_live_sp` if JIT is available, even if the address points to the process memory and no custom allocations were made. Similarly, `EntityPersistentVariable::Dematerialize()` tries to deallocate memory based on the same check, resulting in an error if the memory was not previously allocated in `EntityPersistentVariable::Materialize()`.

As an unintended bonus, the patch also fixes a FIXME case in `TestCxxChar8_t.py`.

igorkudrin committed:
[lldb] Fix evaluating expressions without JIT in an object context

If a server does not support allocating memory in an inferior process or
when debugging a core file, evaluating an expression in the context of a
value object results in an error:

```
error: <lldb wrapper prefix>:43:1: use of undeclared identifier '$__lldb_class'
   43 | $__lldb_class::$__lldb_expr(void *$__lldb_arg)
      | ^
```

Such expressions require a live address to be stored in the value
object. However, `EntityResultVariable::Dematerialize()` only sets
`ret->m_live_sp` if JIT is available, even if the address points to the
process memory and no custom allocations were made. Similarly,
`EntityPersistentVariable::Dematerialize()` tries to deallocate memory
based on the same check, resulting in an error if the memory was not
previously allocated in `EntityPersistentVariable::Materialize()`.

As an unintended bonus, the patch also fixes a FIXME case in
`TestCxxChar8_t.py`.

jimingham commented:
This makes sense.  But the changes that you made are in code that's also used by persistent expression variables - not just persistent results.  If you can't allocate memory, you can't make non-scalar persistent expression variables:

```
(lldb) expr char $my_arr[4] = {'a', 'b', 'c', 'd'}
            
error: Can't evaluate the expression without a running target due to: Interpreter doesn't handle one of the expression's opcodes

```

This is failing in IR interpretation, before it gets to Dematerialization, so your change shouldn't matter for that right now.  

But we do support scalar persistent expression results in core files currently:

(lldb) expr int $my_int = 5
(lldb) expr $my_int * 2
(int) $0 = 10

and that should still work after your change.  I see no reason why it wouldn't, but it would be good to add some  use of persistent expression variables in your test as well as expression result variables to make sure.

igorkudrin committed:
fixup! fix user-defined persistent variables

igorkudrin commented:
> But we do support scalar persistent expression results in core files currently:
> 
> (lldb) expr int $my_int = 5
(lldb) expr $my_int * 2
(int) $0 = 10
> 
> and that should still work after your change. I see no reason why it wouldn't, but it would be good to add some use of persistent expression variables in your test as well as expression result variables to make sure.

Thank you for the test. It revealed another execution path that had to be fixed. I've updated the patch and added your test. Please take a look.

jimingham commented:
That seems like a sensible strategy.

I made one trivial comment about the ordering of parameters, but otherwise this LGTM.

igorkudrin commented on lldb/source/Expression/IRMemoryMap.cpp:
```diff
@@ -321,7 +321,8 @@ IRMemoryMap::Allocation::Allocation(lldb::addr_t process_alloc,
 lldb::addr_t IRMemoryMap::Malloc(size_t size, uint8_t alignment,
                                  uint32_t permissions, AllocationPolicy policy,
-                                 bool zero_memory, Status &error) {
+                                 bool zero_memory, Status &error,
```
That would mean more intrusive code changes because `used_policy` would become a required parameter. Maybe it would make more sense to switch to using `llvm::Expected<>` for this method?

jimingham commented:
It isn't terribly important.  We are switching to returning expected so if you think that looks nicer, feel free to do that.  But otherwise, this is fine as it.

jimingham reviewed (approved):
This one is ready then.

igorkudrin committed:
Merge branch 'main' into expr-result-no-jit
